### PR TITLE
feat: first-run onboarding experience and user engagement nudges

### PR DESCRIPTION
**Pull request by mraduldubey**
## Summary

- **One-time ASCII banner + getting started guide** on first MCP tool call after install
- **Contextual nudges** that fire at the right moment in the user journey (post-registration, post-first-prompt, multi-member milestone) — each shown at most once
- **Welcome-back message** with fleet state on subsequent server startups
- **Upgrade-safe**: existing users skip the banner; corrupt state files recover gracefully

## Architecture

- Central `wrapTool()` in `index.ts` replaces 21 inline wrappers — prepends/appends onboarding text to tool responses
- `src/services/onboarding.ts` — state service (in-memory singleton, atomic file writes, upgrade detection)
- `src/onboarding/text.ts` — all user-facing text constants (banner, guide, nudges, welcome-back)
- State persisted to `~/.apra-fleet/data/onboarding.json` (respects `APRA_FLEET_DATA_DIR`)
- JSON tool responses are never modified (isJsonResponse guard)

## Token cost estimate

- ~370 tokens one-time (banner + guide)
- ~20 tokens/server-start recurring (welcome-back)
- ~80 tokens total nudges across the full user journey

## Test plan

- [x] 57 new tests (658 total), zero regressions
- [x] First call shows banner + guide, second call does not
- [x] JSON tools skip banner prepend
- [x] Nudge sequence: first-register → multi-member → first-prompt (each once)
- [x] Welcome-back: null on first run, shows once per lifecycle
- [x] Upgrade detection: existing registry + no state file = skip banner
- [x] Corruption recovery: malformed JSON = default state
- [x] Atomic writes (temp + rename)
- [x] formatLastActive NaN guard on malformed timestamps

## Review history

- Plan: 2 cycles (6 must-fix items → all resolved → APPROVED)
- Phase 1 (Foundations): APPROVED
- Phase 2 (Core Integration — riskiest): APPROVED, zero regressions
- Phase 3 (Contextual Nudges): APPROVED
- Phase 4 (Hardening): APPROVED — all 5 non-blocking RECs resolved

### CODE REVIEW

**Comment by kumaakh** [comment]
Interesting ideas, will review

**Comment by mraduldubey** [comment]
> Interesting ideas, will review

There is a issue where the claude code AI "surpresses" or paraphrases the onboarding text and nudged. WIP

File: feedback.md [hunk 1]
```diff
@@ -0,0 +1,85 @@
+# Onboarding Banner Visibility Fix — Code Review
```
Comment (kumaakh) [inline feedback.md]: **Stale file: should not be committed.** This looks like a review artifact from a prior round of feedback. It doesn't belong in the repo — it'll confuse future contributors who see it alongside the ADR. Please remove before merge.

File: .gitignore [hunk 2]
```diff
@@ -10,3 +10,4 @@ fleet.config.json
 ~/.claude-fleet/
 materials/*
 .claude/settings.local.json
+CLAUDE.md
```
Comment (kumaakh) [inline .gitignore:13]: **Noted: CLAUDE.md now gitignored.** This means the project-level CLAUDE.md (which contains tool reference, workflows, and installation instructions) will no longer be tracked. If this is intentional (e.g., each contributor maintains their own), that's fine — but it means new clones won't have it. Consider whether the project-level instructions should remain tracked while only a local override is ignored.

**Review by kumaakh — DISMISSED** [review]

File: src/index.ts [hunk 3]
```diff
@@ -71,47 +79,109 @@ async function startServer() {
...
+    } catch (e: unknown) {
+      const msg = (e instanceof Error ? e.message : String(e));
+      if (!/logging|method not found|not supported/i.test(msg)) {
+        process.stderr.write(`[apra-fleet] onboarding notification failed: ${msg}\n`);
+      }
+    }
+  }
+
+  function sanitizeToolResult(s: string): string {
+    return s.replace(/<\/?apra-fleet-display[^>]*>/gi, '[tag-stripped]');
```
Comment (kumaakh) [inline src/index.ts]: ** — unterminated tag edge case**

The regex `/<\/?apra-fleet-display[^>]*>/gi` correctly handles attributes and case variants (covered by tests at line 820). However, it requires a closing `>` — a deliberately malformed tag like `<apra-fleet-display ` (no closing bracket) would pass through unstripped.

In practice this is low-severity because: (1) tool result text comes from our own code, not arbitrary external input, and (2) the MCP client still has to interpret the unclosed tag as meaningful. Worth adding a test case and optionally a more defensive pattern like `/<apra-fleet-display[\s\S]*?(?:>|$)/gi` if threat model includes adversarial tool output.

File: src/tools/register-member.ts [hunk 4]
```diff
@@ -19,13 +19,13 @@ export const registerMemberSchema = z.object({
     .regex(/^[a-zA-Z0-9._-]+$/, 'Only letters, numbers, dots, dashes, and underscores')
     .describe('Human-friendly name for this member (worker) (e.g. "web-server")'),
   member_type: z.enum(['local', 'remote']).default('remote').describe('Member type: "local" for same machine, "remote" for SSH (default: "remote")'),
-  host: z.string().optional().describe('IP address or hostname of the remote machine (required for non-cloud remote members; optional for cloud members — auto-resolved from AWS when running)'),
+  host: z.string().regex(/^[^<>\n\r]+$/, 'host must not contain angle brackets or newlines').optional().describe('IP address or hostname of the remote machine (required for non-cloud remote members; optional for cloud members — auto-resolved from AWS when running)'),
```
Comment (kumaakh) [inline src/tools/register-member.ts:22]: **update_member missing host/work_folder validation — exploitable gap**

register_member adds regex guards on host (`/^[^\<\>\n\r]+$/`) and work_folder to block angle-bracket injection. update_member has no equivalent guards. An LLM or caller can update an existing member's host to a value containing `<` or `>` (e.g. to inject fake `<apra-fleet-display>` content into a subsequent execute_command result).

This is not purely theoretical: since sanitizeToolResult only strips tags in the main tool result string, a crafted host/folder value that ends up in the result of a status command would pass through. Recommend adding the same Zod .regex() validators to update_member's host and work_folder fields.

File: src/index.ts [hunk 5]
```diff
@@ -71,47 +79,109 @@ async function startServer() {
...
-  server.tool('list_members', 'List all fleet members and their current status. Use format="json" for structured data.', listMembersSchema.shape, async (input) => ({ content: [{ type: 'text', text: await listMembers(input as any) }] }));
-  server.tool('remove_member', 'Remove a member from the fleet.', removeMemberSchema.shape, async (input) => ({ content: [{ type: 'text', text: await removeMember(input as any) }] }));
-  server.tool('update_member', "Change a member's name, connection details, working directory, AI provider, or other settings.", updateMemberSchema.shape, async (input) => ({ content: [{ type: 'text', text: await updateMember(input as any) }] }));
+  server.tool('register_member', 'Add a machine to the fleet. Use member_type "local" for this machine or "remote" for a machine reachable over SSH. Choose the AI provider the member will use for prompts.', registerMemberSchema.shape, wrapTool('register_member', (input) => registerMember(input as any)));
```
Comment (kumaakh) [inline src/index.ts:149]: **wrapTool coverage confirmed — all 21 tools wrapped**

Counted all server.tool() registrations: register_member, list_members, remove_member, update_member, send_files, receive_files, execute_prompt, execute_command, provision_llm_auth, setup_ssh_key, setup_git_app, provision_vcs_auth, revoke_vcs_auth, fleet_status, member_detail, update_llm_cli, shutdown_server, version, compose_permissions, cloud_control, monitor_task. All 21 are wrapped via wrapTool(). No handler is registered bare. Good.

File: src/onboarding/text.ts [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+export function WELCOME_BACK(memberCount: number, lastActive: string): string {
+  if (memberCount === 0) {
+    return '── Apra Fleet ──────────────────────────────────────\nFleet ready. Register a member to get started.\n────────────────────────────────────────────────────';
+  }
+  const plural = memberCount !== 1 ? 's' : '';
+  return `── Apra Fleet ──────────────────────────────────────\nFleet: ${memberCount} member${plural} · Last active: ${lastActive}\n────────────────────────────────────────────────────`;
+}
+
+/**
+ * Nudge shown after the user registers their first member.
```
Comment (kumaakh) [inline src/onboarding/text.ts:102]: The `NUDGE_AFTER_FIRST_REGISTER` padding uses `Math.max(1, 28 - memberName.length)` — if `friendly_name` is longer than ~27 chars the tip box's right border will misalign. The schema regex allows unlimited length. Consider adding a `.max(30)` or similar to the `friendly_name` Zod schema, or truncating the name in the nudge display. Non-blocking.

File: src/onboarding/text.ts [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+ */
+export function NUDGE_AFTER_MULTI_MEMBER(): string {
+  return `\n┌─ Tip ──────────────────────────────────────────────────────┐\n│ 🤝 You have multiple members — try the PM skill:           │\n│    /pm init  →  /pm pair  →  /pm plan                      │\n│    One member builds, another reviews — across machines.   │\n└────────────────────────────────────────────────────────────┘`;
+}
+
+/**
+ * Server-level instruction injected once at `initialize` time.
+ * Instructs the LLM to reproduce verbatim any text inside <apra-fleet-display> tags.
+ */
+export const VERBATIM_INSTRUCTIONS = `When a tool response contains text wrapped in \`<apra-fleet-display>…</apra-fleet-display>\` tags, that content is user-facing onboarding material. Reproduce it VERBATIM in your reply to the user as the very first thing you output, preserving all formatting, line breaks, and emoji. Do not paraphrase, summarize, translate, or comment on it. Do not wrap it in code fences. Strip the \`<apra-fleet-display>\` tags themselves — they are for your recognition only.`;
```
Comment (kumaakh) [inline src/onboarding/text.ts:132]: **VERBATIM_INSTRUCTIONS — achieves ADR intent well**

The instruction text is clear and actionable: it specifies the tag name, directs the LLM to output content verbatim as the first thing in the reply, preserve formatting/emoji, strip the tags, and explicitly forbids paraphrase/summarize/code-fences. This directly addresses the ADR problem statement (LLMs collapsing/paraphrasing onboarding UX).

One note: the instruction says 'as the very first thing you output' — this may conflict with LLMs that prepend a thinking preamble or are in extended thinking mode. Not a blocker, but worth noting in the ADR as a known limitation.

File: src/services/onboarding.ts [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+/**
+ * Persist the current in-memory state to disk atomically (temp write + rename).
+ */
+export function saveOnboardingState(): void {
+  if (_state === null) return;
+  ensureFleetDir();
+
+  const tmp = ONBOARDING_PATH + '.tmp';
+  fs.writeFileSync(tmp, JSON.stringify(_state, null, 2), { mode: 0o600 });
+  fs.renameSync(tmp, ONBOARDING_PATH);
```
Comment (kumaakh) [inline src/services/onboarding.ts:80]: **Atomic write pattern is correct**

The temp-file + rename pattern (writeFileSync to .tmp, renameSync to final path, enforceOwnerOnly after) is the correct approach for atomic writes on POSIX systems. A crash between write and rename leaves only the .tmp orphan — the original file is untouched. enforceOwnerOnly is correctly called on the final path after rename, not on the temp file.

Minor: enforceOwnerOnly is called after renameSync — this creates a brief window (between rename and chmod) where the file exists at the final path with potentially wrong permissions. If onboarding.json is created fresh (no pre-existing file), the mode is already 0o600 from writeFileSync, so the window is harmless. But if the system umask is permissive and writeFileSync ignores the mode on some platforms, the window is real. Low severity.

**Review by kumaakh — DISMISSED** [review]

**Review by kumaakh — DISMISSED** [review]
superseded

File: feedback.md (hunk 1 above)
Comment (kumaakh) [reply feedback.md]: /pm cleanup already gets rid of these files... it is not made automated so that human can review it before calling the sprint done

File: .gitignore (hunk 2 above)
Comment (kumaakh) [inline .gitignore]: question: Adding CLAUDE.md to .gitignore means future changes to the checked-in project instructions won't be tracked. Is this intentional (to allow local overrides)? If so, worth a brief comment.

**Review by kumaakh — DISMISSED** [review]

**Review by kumaakh — DISMISSED** [review]

**Review by kumaakh — DISMISSED** [review]

**Review by kumaakh — DISMISSED** [review]

File: src/index.ts (hunk 3 above)
Comment (kumaakh) [inline src/index.ts]: **Regex is solid.** `/<\/?apra-fleet-display[^>]*>/gi` correctly handles: case variants, attributes (`foo="x"`), self-closing, and multiple occurrences. The `[^>]*` character class also spans newlines in JS, so attribute-based bypass attempts won't work.

One edge case: a tag without a closing `>` (e.g., `<apra-fleet-display `) won't match, but that's benign — an unclosed tag isn't valid markup. No action needed.

**Note:** `update_member` schema (`src/tools/update-member.ts:17,27`) does NOT have the `host`/`work_folder` angle-bracket regex that `register_member` added. An attacker could register a clean member then `update_member` with `host: "1.2.3.4<apra-fleet-display>injected</apra-fleet-display>"` to bypass schema-level defense. `sanitizeToolResult` catches it at output, so blast radius is limited — but defense-in-depth should apply the same regex there. Consider extracting a shared `safeString` Zod refinement.

File: src/index.ts (hunk 5 above)
Comment (kumaakh) [inline src/index.ts]: **wrapTool coverage confirmed: all 21 tools wrapped**

Verified every `server.tool()` registration (lines 149-183) passes its handler through `wrapTool`. No tool bypasses sanitization or the onboarding preamble pipeline. Complete coverage. 👍

File: src/services/onboarding.ts (hunk 8 above)
Comment (kumaakh) [inline src/services/onboarding.ts]: **Atomic write pattern: correct implementation**

The temp-file + rename pattern is textbook POSIX atomic write: `writeFileSync` to `.tmp` with restrictive `mode: 0o600`, then `renameSync` (atomic on same filesystem), then `enforceOwnerOnly` for belt-and-suspenders permission enforcement.

Minor note: if `writeFileSync` throws mid-write (disk full), the `.tmp` file is left as a stale artifact. It won't affect correctness since the main file is untouched, and for a config file that changes rarely, cleanup complexity isn't worth adding. Test at onboarding.test.ts:118-126 correctly verifies no temp file remains on the happy path.

**Review by kumaakh — DISMISSED** [review]
superseded

**Review by kumaakh — DISMISSED** [review]
superseded

File: src/onboarding/text.ts (hunk 7 above)
Comment (kumaakh) [inline src/onboarding/text.ts]: **VERBATIM_INSTRUCTIONS: effective, with one phrasing nit**

Clear and specific — "reproduce VERBATIM", "strip the tags", "do not paraphrase". The multi-channel delivery (notifications + content blocks + audience annotation) is smart defense-in-depth: even if the LLM ignores the instruction, the client rendering layer gets a second chance via annotations.

Nit: "as the very first thing you output" could be slightly misleading if the LLM has other system-level instructions that also request being first. In practice this works fine since MCP tool results arrive before the LLM generates its reply, so there's no real conflict. Just noting for future reference.

File: src/services/onboarding.ts [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+  enforceOwnerOnly(ONBOARDING_PATH);
```
Comment (kumaakh) [inline src/services/onboarding.ts]: **Atomic write pattern: correct ✅**

`writeFileSync` to `.tmp` with mode `0o600`, then `renameSync` (atomic on POSIX), then `enforceOwnerOnly` as belt-and-suspenders. If the process crashes between `renameSync` and `enforceOwnerOnly`, the file already has 0o600 from the initial write — no permission window. Tests verify no temp file left behind (line 118 of test file). Solid.

**Review by kumaakh — DISMISSED** [review]
superseded

File: src/onboarding/text.ts (hunk 7 above)
Comment (kumaakh) [inline src/onboarding/text.ts:132]: **VERBATIM_INSTRUCTIONS achieves what the ADR claims — within inherent limits.** The instruction is clear and uses specific verbs ("reproduce", "strip", "do not paraphrase"). The three-channel delivery means any single channel reaching the user suffices. The notification channel (Channel 1) provides a deterministic fallback that bypasses the LLM entirely, so even if the LLM ignores this instruction, the user still sees the text.

This is about as good as you can get without client-side changes. Well designed.

File: src/tools/register-member.ts (hunk 4 above)
Comment (kumaakh) [inline src/tools/register-member.ts:22]: **Schema validation is a nice defense-in-depth layer.** The `host` and `work_folder` regexes (`/^[^<>\n\r]+$/`) block angle brackets and newlines at the input boundary — clean and minimal. This prevents marker injection before it reaches `sanitizeToolResult`.

One thing: `friendly_name` uses a strict allowlist (`/^[a-zA-Z0-9._-]+$/`) which is inherently safe. The asymmetry between `friendly_name` (allowlist) and `host`/`work_folder` (denylist) is reasonable given these fields need different character sets.

**Review by kumaakh — DISMISSED** [review]
superseded

**Review by kumaakh — CHANGES_REQUESTED** [review]
## Review — feat/onboarding-ux

### Test Results

| Check | Result |
|-------|--------|
| `npm test` | 717 passed, 4 skipped, 1 failed (crypto timeout — unrelated to this PR) |
| All onboarding tests | ✅ Pass (888 lines of coverage) |
| TypeScript compilation | Clean |

### Summary

Three-channel onboarding delivery is well-designed, test coverage is comprehensive, atomic writes correct, wrapTool covers all 21 tools. One blocker before merge.

### Blocking

**update_member missing host/work_folder validation** — `register_member` got angle-bracket injection guards in this PR but `update_member` did not. A caller can register a clean member then update `host`/`work_folder` to inject `<apra-fleet-display>` content, bypassing the input layer defence. Fix: add the same Zod `.regex()` to `update_member`'s schema — one-liner.

### Non-blocking

1. `sanitizeToolResult` — regex requires closing `>`; unterminated tags pass through. Low severity, worth a test.
2. `feedback.md` committed — review artefact, remove before merge.
3. `CLAUDE.md` gitignored — verify intentional.
4. Tip box padding overflows if `friendly_name` > ~27 chars — consider truncation.

Fix the blocker, then ready to merge.

**Review by kumaakh — APPROVED** [review]
**APPROVED** — thorough review of all 12 changed files (1929 additions, 29 deletions).

**Architecture:** Three-channel delivery (notifications, markers+instructions, audience annotations) is well-designed defense-in-depth. The ADR is excellent — honest about tradeoffs, token costs, and known gaps.

**Code quality:** Clean. `wrapTool` centralizes 21 inline wrappers. State management is solid (atomic writes, corruption recovery, upgrade detection, forward-compatible defaults). Sanitization handles edge cases (case variants, attributes, unterminated tags).

**Security:** Two-layer injection defense (output sanitization + input validation) is correct. Both `register_member` and `update_member` now validate against angle brackets.

**Testing:** 722 tests pass, zero failures. 57 new onboarding tests + 21 text tests + smoke test cover: fresh install, upgrade path, corruption, milestone progression, idempotency, passive-tool guard, JSON bypass, full session sequence, notification emission, sanitization, schema validation.

**Non-blocking note:** `.gitignore` adds `CLAUDE.md` — no immediate effect since it's already tracked, but looks like a dev artifact. Can clean up in follow-up.
